### PR TITLE
feat: threshold of 1.11m on `equal_shape_distance_diff_coordinates`

### LINKED ISSUE #1638: Test setting threshold of 1.11m on equal_shape_dist_diff_coordinates
### Describe the problem

There are recurring concerns that the [equal_shape_dist_diff_coordinates error is too sensitive](https://github.com/MobilityData/gtfs-validator/issues/1258) to very small differences between coordinate values. Upon looking at the history of the rule, it was originally proposed to [use 1.11m as a threshold between coordinates](https://github.com/MobilityData/gtfs-validator/pull/1083#issuecomment-1004250458). There was discussion to [set it as an ERROR above 1.11m and a WARNING below 1.11m](https://github.com/MobilityData/gtfs-validator/pull/1083#issuecomment-1010188909). At the time when we had fewer feeds in the Mobility Database, this seemed to have minimal value (e.g few feeds were impacted by the threshold change). Now that it's two years later, it's worth testing this with our current feed repository as a starting point to see how many feeds are impacted by it, and propose a more permissive threshold if needed. 

### Proposed solution

- Apply a 1.11m threshold to equal_shape_dist_diff_coordinates ERROR (>= 1.11m)
- Create a WARNING for equal_shape_dist_diff_coordinates( <1.11m)
- Re-run the acceptance tests to see how many feeds are impacted and what the difference is between coordinates for the feeds affected. 

### Alternatives you've considered

- [Defining a threshold based on analysis run on percentiles from the Mobility Database](https://github.com/MobilityData/gtfs-validator/issues/1258). Using 1.11m as a threshold seems like a good starting point to test the impact of values on the feeds.

### Additional context

Current rules that have set thresholds, to review logic: 

- [fast_travel_between_consecutive_stops](https://gtfs-validator.mobilitydata.org/rules.html#fast_travel_between_consecutive_stops-rule)
- [fast_travel_between_far_stops](https://gtfs-validator.mobilitydata.org/rules.html#fast_travel_between_far_stops-rule)
- [stop_too_far_from_shape](https://gtfs-validator.mobilitydata.org/rules.html#stop_too_far_from_shape-rule)

### LINKED ISSUE #1678: Fix tooltip for actualDistanceBetweenShapePoints
### Describe the bug

It looks like there's a [typo in the markdown for the description of actualDistanceBetweenShapePoints](https://github.com/MobilityData/gtfs-validator/blob/69b948e7f96ce78ea45e0932a1736520c7387334/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java#L214) that is causing only the last word of the sentence to be displayed.

### Steps/Code to Reproduce

Screenshot 2024-02-21 at 10 52 47 AM

[Test feed to run](https://github.com/MobilityData/gtfs-validator/files/14361840/1882.zip)

- Run validator and open the equal_shape_distance_diff_coordinates notice

### Expected Results

Full sentence displays. Please update sentence to "Actual distance traveled along the shape from the first shape point to the previous shape point. Distance units defined in the feed." 

### Actual Results

See above in Steps to reproduce.

### Screenshots

_No response_

### Files used

_No response_

### Validator version

4.2

### Operating system

MacOS

### Java version

_No response_

### Additional notes

_No response_

### PR DESCRIPTION
**Pull request by cka-y**
**Summary:**
Closes #1638 
Closes #1678 
Concerns have been raised about the `equal_shape_dist_diff_coordinates` error's sensitivity to minor coordinate discrepancies. [Historical discussions](https://github.com/MobilityData/gtfs-validator/pull/1675#issuecomment-1951619648) proposed a $1.11 m$ threshold to distinguish between `ERROR` ($\ge 1.11m$) and `WARNING` ($\lt 1.11 m$) levels. Given the expansion of the Mobility Database since these initial discussions, reevaluating this threshold with current data offers a chance to refine the rule's sensitivity.

**Expected behavior:** 
- Introduces a 1.11m threshold for `equal_shape_distance_diff_coordinates`, triggering an `ERROR` for distances $\geq 1.11m$.
- Creates a new notice, `equal_shape_distance_diff_coordinates_below_threshold`, with `WARNING` severity for distances $<1.11m$.

**Acceptance tests results:**
The acceptance tests [results](https://github.com/MobilityData/gtfs-validator/pull/1675#issuecomment-1952912555) show a 6% reduction (93 notices) in `ERROR` notices. Additionally, 166 new `WARNING` notices were generated. This is because certain datasets exhibited distances between coordinates that were both greater than and less than $1.11m$, thus qualifying for both `WARNING` and `ERROR` notices.

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [ ] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Comment by emmambd** [comment]
[I did a deep dive into the data reported during our last release (4.2)](https://docs.google.com/spreadsheets/d/1It3g1O3TGCLnE8uR3x-ThKErwgh3TTjm-j-ABwiaANk/edit#gid=0) and used that as the source of truth to compare the original number of feeds that triggered this error vs. the revised list from these acceptance tests.

This threshold reduces the number of feeds triggered by this error by 50% (174 -> 87).

If there are persistent user concerns after this rule change, we can revisit the threshold by doing analysis on all notice occurrences with the updated rule and the difference in metres for each occurrence, and then doing a distribution graph to help determine a more lenient threshold. 

For this release iteration, 1.11m is sufficient.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 1]
```diff
@@ -72,8 +74,15 @@ public void validate(NoticeContainer noticeContainer) {
...
-          noticeContainer.addValidationNotice(
-              new EqualShapeDistanceDiffCoordinatesNotice(prev, curr));
+          double distanceBetweenShapePoints =
+              getDistanceMeters(curr.shapePtLatLon(), prev.shapePtLatLon());
+          if (distanceBetweenShapePoints >= DISTANCE_THRESHOLD) {
+            noticeContainer.addValidationNotice(
+                new EqualShapeDistanceDiffCoordinatesNotice(prev, curr));
+          } else if (distanceBetweenShapePoints > 0) {
+            noticeContainer.addValidationNotice(
+                new EqualShapeDistanceDiffCoordinatesWarningNotice(prev, curr));
```
Comment (emmambd) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:84]: Update warning notice name to: `equal_shape_distance_diff_coordinates_distance_below_1.1m`

**Review by emmambd — COMMENTED** [review]
@cka-y 1 minor comment re: notice name

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 1 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:84]: i cannot put a `.` in the name of the notice

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 2]
```diff
@@ -225,4 +234,58 @@ static class EqualShapeDistanceDiffCoordinatesNotice extends ValidationNotice {
...
+    private final long prevCsvRowNumber;
+
+    /** The previous shape point's `shape_dist_traveled` value. */
+    private final double prevShapeDistTraveled;
+
+    /** The previous record's `shapes.shape_pt_sequence`. */
+    private final int prevShapePtSequence;
+
+    /**
+     * Actual distance traveled along the shape from the first shape point to the previous shape.
```
Comment (emmambd) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:274]: @cka-y I just realized since we generate documentation automatically now that #1678 is easy to solve in this PR (fixing the typo here).

Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java]: ✅ done

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 3]
```diff
@@ -72,8 +74,15 @@ public void validate(NoticeContainer noticeContainer) {
         }
         // equal shape_dist_traveled and different coordinates
         if (!(curr.shapePtLon() == prev.shapePtLon() && curr.shapePtLat() == prev.shapePtLat())) {
-          noticeContainer.addValidationNotice(
-              new EqualShapeDistanceDiffCoordinatesNotice(prev, curr));
+          double distanceBetweenShapePoints =
+              getDistanceMeters(curr.shapePtLatLon(), prev.shapePtLatLon());
+          if (distanceBetweenShapePoints >= DISTANCE_THRESHOLD) {
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:79]: We have to keep in mind that the distance calculated from coordinates depends on the latitude. 
So a difference of .00001 in longitude gives a distance of 1.11 m at the equator, and .8 m in Montreal.
that means that this theoretical case would give an error at the equator and a warning in Montreal.
It's an edge case, but maybe we should test for `>` instead if `>=` ?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 4]
```diff
@@ -225,4 +236,59 @@ static class EqualShapeDistanceDiffCoordinatesNotice extends ValidationNotice {
           getDistanceMeters(current.shapePtLatLon(), previous.shapePtLatLon());
     }
   }
+
+  /**
+   * Two consecutive points have equal `shape_dist_traveled` and different lat/lon coordinates in
+   * `shapes.txt` and the distance between the two points is less than 1.11m.
+   *
+   * <p>When sorted by `shape.shape_pt_sequence`, the values for `shape_dist_traveled` must increase
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:244]: Why is an error mentioned when it's a warning? Shouldn't the 1.11 m be mentioned somewhere?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 5]
```diff
@@ -225,4 +236,59 @@ static class EqualShapeDistanceDiffCoordinatesNotice extends ValidationNotice {
           getDistanceMeters(current.shapePtLatLon(), previous.shapePtLatLon());
     }
   }
+
+  /**
+   * Two consecutive points have equal `shape_dist_traveled` and different lat/lon coordinates in
+   * `shapes.txt` and the distance between the two points is less than 1.11m.
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:242]:
```suggestion
   * `shapes.txt` and the distance between the two points is less more than 0 and than 1.11m.
```

Comment (emmambd) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:242]: @jcpitre do you mean "is more than 0 and less than 1.11m"?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 6]
```diff
@@ -225,4 +236,59 @@ static class EqualShapeDistanceDiffCoordinatesNotice extends ValidationNotice {
...
+   *
+   * <p>When sorted by `shape.shape_pt_sequence`, the values for `shape_dist_traveled` must increase
+   * along a shape. Two consecutive points with equal values for `shape_dist_traveled` and different
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:245]:
```suggestion
   * along a shape. Two consecutive points with equal values for `shape_dist_traveled` and small difference of coordinates (less than 1.11 m) result in a warning.
```

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 3 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:79]: i'd have to change the name of the notice as well to `...bellow_or_equal_to_threshold`. Is this a change we want?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 5 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:242]: isn't a distance always greater than 0 anyway?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 3 above)
Comment (jcpitre) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:79]: Fine. The operative word was `maybe`

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 5 above)
Comment (jcpitre) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:242]: No, but I guess people will understand.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java [hunk 7]
```diff
@@ -225,4 +236,59 @@ static class EqualShapeDistanceDiffCoordinatesNotice extends ValidationNotice {
...
     }
   }
+
+  /**
+   * Two consecutive points have equal `shape_dist_traveled` and different lat/lon coordinates in
+   * `shapes.txt` and the distance between the two points is less than 1.11m.
+   *
+   * <p>When sorted by `shape.shape_pt_sequence`, the values for `shape_dist_traveled` must increase
+   * along a shape. Two consecutive points with equal values for `shape_dist_traveled` and small difference of coordinates (less than 1.11 m distance) result in a warning.
+   * coordinates indicate an error.
```
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java:246]:
```suggestion
```

**Review by jcpitre — APPROVED** [review]

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java (hunk 7 above)
Comment (cka-y) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/ShapeIncreasingDistanceValidator.java]: fixed it on my branch ✅
